control_panel_api/tests/test_views.py

```diff
@@ -224,6 +225,37 @@ def test_update(self):
         self.assertEqual(HTTP_200_OK, response.status_code)
         self.assertEqual(data['access_level'], response.data['access_level'])
 
+    @patch('control_panel_api.models.AppS3Bucket.update_aws_permissions')
+    def test_update_updates_aws(self, mock_update_aws_permissions):
+        data = {
+            'app': self.app_1.id,
+            's3bucket': self.s3_bucket_1.id,
+            'access_level': AppS3Bucket.READWRITE,
+        }
+        self.client.put(
+            reverse('apps3bucket-detail', (self.apps3bucket_1.id,)), data)
+        mock_update_aws_permissions.assert_called()
+
+    def test_update_when_app_changed(self):
+        data = {
+            'app': self.app_2.id,
+            's3bucket': self.s3_bucket_1.id,
+            'access_level': AppS3Bucket.READWRITE,
+        }
+        response = self.client.put(
+            reverse('apps3bucket-detail', (self.apps3bucket_1.id,)), data)
+        self.assertEqual(HTTP_400_BAD_REQUEST, response.status_code)
+
+    def test_update_when_s3bucket_changed(self):
+        data = {
+            'app': self.app_1.id,
+            's3bucket': self.s3_bucket_2.id,
+            'access_level': AppS3Bucket.READWRITE,
+        }
+        response = self.client.put(
+            reverse('apps3bucket-detail', (self.apps3bucket_1.id,)), data)
+        self.assertEqual(HTTP_400_BAD_REQUEST, response.status_code)
```

Comment:
These last two tests look very similar, can they be data driven? Like parametrized tests with pytest? Eg:
```python
def test_updates(self):
    cases = [
        {'app': self.app_2.id, 's3bucket': self.s3_bucket_1.id, ...},
        {'app': self.app_1.id, 's3bucket': self.s3_bucket_2.id, ...}
    ]
    for case in cases:
        self.assert_update_fails(self.apps3bucket_1, case)
```

Reply:
These tests have some common lines of code, which could be made clearer with a custom assert method, eg:
```python
def test_update_when_s3bucket_changed(self):
    data = {...}
    self.assert_update_fails(self.apps3bucket_1, data)
```